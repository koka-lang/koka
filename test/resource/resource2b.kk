// Test resource handler

effect filesys<s::S> { }

fun filesys( action : forall<s> () -> <filesys<s>|e> a ) : e a {
  handle<filesys<_>>(action){ }
} 

effect resource file<s::S> in filesys<s> {
  fun read() : string
}


fun fake-file(content,action) {
  handle resource (action) {
    return x -> x
    read() -> resume(content)
  }
}

fun main() {
  using filesys
  use f = fake-file( "hi" ) 
  println(f.read.count + 40)
}