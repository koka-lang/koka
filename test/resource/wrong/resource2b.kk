// Test scoped handlers

effect heap<s::S> { 
  fun get-ref() : ref<s>
}

type ref<s::S> {
  Ref
}

fun get( r : ref<s> ) : heap<s> int {
  42
}

fun heap( action : forall<s> () -> <heap<s>|e> a) : e a {
  handle(action) { 
    get-ref() -> resume(Ref)
  }
}

fun foo() {
  using heap
  get-ref()
}

fun bar() {
  using heap
  foo().get   // wrong: trying to access reference out of scope
}

fun main() {
  println(bar())
}
