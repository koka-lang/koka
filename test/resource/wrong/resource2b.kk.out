test/resource/wrong/resource2b.kk(22, 3): error: type is not polymorph enough
 context : using heap
 term : using heap
 get-ref()
 inferred type : forall<s1> () -> (heap<s1>) ref<s1>
 is less polymorph as: forall<s1> () -> <heap<s1>|_e> ref<$s>